File: Haskell/004.hs
{-
  A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.

  Find the largest palindrome made from the product of two 3-digit numbers.
-}

import Util (isPalindrome)

threeDigitProductPermutation :: [Integer]
threeDigitProductPermutation = [x * y | x <- z, y <- z]
  where z = [100..999]

main :: IO ()
main = print . maximum . map (\x -> read x :: Integer) . filter isPalindrome . map show $ threeDigitProductPermutation

File: Python/004.py
'''
A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.

Find the largest palindrome made from the product of two 3-digit numbers.
'''

def isPalindrome(x):
	return str(x) == str(x)[::-1]

def threeDigitProductPermutation():
	threeDigitRange = range(100, 999)
	return [i * j for i in threeDigitRange for j in threeDigitRange]

print(max(list(filter(isPalindrome, threeDigitProductPermutation()))))
